revised SDL-0280 (#997)
smartdevicelink/sdl_evolution: proposals/0280-Adding-new-parameter-of-requiresAudioSupport-and-BluetoothDeviceAddress.md

# Adding new parameter of requiresAudioSupport and BluetoothDeviceAddress

* Proposal: [SDL-0280](0280-Adding-new-parameter-of-requiresAudioSupport-and-BluetoothDeviceAddress.md)
* Author: [Shohei Kawano](https://github.com/Shohei-Kawano), [Kazuki Sugimoto](https://github.com/Kazuki-Sugimoto)
* Status: **In Review**

> * Impacted Platforms: [ Core / iOS / Java Suite / RPC / HMI ]

Reviewer: I believe `Protocol` needs to be added here, based on the requested revisions.

### Change the app registration flow

In order for the app to be always registered, the changes in SDL session establishment to app registration flow are shown below.
If the SDL app or HU cannot use this function, it will perform the same operation as before.

![registeringApp_flow](../assets/proposals/0280-Adding-new-parameter-of-requiresAudioSupport-and-BluetoothDeviceAddress/registeringApp_flow1.png)

1. The developer will set their `requiresAudioSupport` to true.
2. The library will send a `StartService` for the RPC service with a new param in the payload, `requiresAudioSupport`.
3. Core will receive the `StartService` for the RPC service: 

>     i. Core will check it's current audio connects (BT A2DP), if it has a supported audio connect it will continue in flow.

Reviewer: ```suggestion
    i. Core will check its current audio connects (BT A2DP), if it has a supported audio connect it will continue in flow.
```

>     ii. If there is no Audio support, core will check if it supports the auto connect BT feature. If it does it will continue in flow.

Reviewer: ```suggestion
    ii. If there is no audio support, Core will check if it supports the auto connect BT feature. If it does, it will continue in flow.
```

>     iii. If there is no supported audio methods and Core does not support the auto connect BT feature, it will send a `StartServiceNAK` with the reason `No audio support available`.

Reviewer: Please add line breaks to fix the formatting, so each item in the list is on its own line.

Author: Added line breaks.

4. If Core has continued, it will send a `StartServiceACK` with a new param `autoBTCapability` set to true.
5. The app receives the response to its `StartService` for the RPC service:

>     1.If the response was a `StartServiceNAK` the app will shutdown.

Reviewer: ```suggestion
    1. If the response was a `StartServiceNAK` the app will shut down.
```

>     2. If the response was a `StartServiceACK`, `requiresAudioSupport` was set to true, but the protocol version of the ACK is less than the major version of this feature, the app will shutdown.

Reviewer: ```suggestion
    2. If the response was a `StartServiceACK`, `requiresAudioSupport` was set to true, but the protocol version of the ACK is less than the major version of this feature, the app will shut down.
```

>     3. If the response was a `StartServiceACK`, `requiresAudioSupport` was set to true, the protocol version of the ACK is equal or greater than the major version of this feature, and the `autoBTCapability` flag is set to false, the app will check if audio support is available using the `MediaStreamingStatus` class. If it is available it will continue, if not it will shutdown.

Reviewer: ```suggestion
    3. If the response was a `StartServiceACK`, `requiresAudioSupport` was set to true, the protocol version of the ACK is equal or greater than the major version of this feature, and the `autoBTCapability` flag is set to false, the app will check if audio support is available using the `MediaStreamingStatus` class. If it is available it will continue, if not it will shut down.
```

>     4. If the response was a `StartServiceACK`, `requiresAudioSupport` was set to true, the protocol version of the ACK is equal or greater than the major version of this feature, and the `autoBTCapability` flag is set to true, the app will continue.

Reviewer: See attached screenshot of current formatting.  Please add line breaks to fix the formatting, so each item in the list is on its own line.  These should also be Roman Numerals to be consistent with the other sub-bullets.
<img width="945" alt="Screen Shot 2020-05-18 at 8 15 36 PM" src="https://user-images.githubusercontent.com/22600496/82271124-ec546080-9944-11ea-8036-1f3b53350b2f.png">

Author: Adapted to the format.

6. The app will send its `RegisterAppInterface` which will include the `hmiTypes` and `isMediaApplication` flag.
7. Core will receive the `RegisterAppInterface`: 

>     i. If the `requiresAudioSupport` flag was not included in the `StartService`, `isMediaApplication` is set to true in the RAI and the protocol version for the session is less than the major version of the version that included this feature it will send a RegisterAppInterface response with success=false and deny the app's registration.

Reviewer: ```suggestion
    i. If the `requiresAudioSupport` flag was not included in the `StartService`, `isMediaApplication` is set to true in the RAI and the protocol version for the session is less than the major version of the version that included this feature, it will send a `RegisterAppInterface` response with `success=false` and deny the app's registration.
```

>     ii. If the `requiresAudioSupport` flag was not included in the `StartService`, `isMediaApplication` is set to true in the RAI and the protocol version for the session is equal to or greater than the major version of the version that included this feature it will send a RegisterAppInterface response with `success=true` but not move forward with this proposals feature.

Reviewer: ```suggestion
    ii. If the `requiresAudioSupport` flag was not included in the `StartService`, `isMediaApplication` is set to true in the RAI and the protocol version for the session is equal to or greater than the major version of the version that included this feature, it will send a `RegisterAppInterface response` with `success=true` but not move forward with this proposal's feature.
```

>     iii. If the `requiresAudioSupport` flag was set to true in the `StartService`, `isMediaApplication` is set to true in the RAI and the protocol version for the session is equal to or greater than the major version of the version that included this feature it will send a RegisterAppInterface response with `success=true` and move forward with this proposals feature.

Reviewer: Please add line breaks to fix the formatting, so each item in the list is on its own line.

Author: Added line breaks.

>     iii. If the `requiresAudioSupport` flag was set to true in the `StartService`, `isMediaApplication` is set to true in the RAI and the protocol version for the session is equal to or greater than the major version of the version that included this feature it will send a RegisterAppInterface response with `success=true` and move forward with this proposals feature.

Reviewer: ```suggestion
    iii. If the `requiresAudioSupport` flag was set to true in the `StartService`, `isMediaApplication` is set to true in the RAI and the protocol version for the session is equal to or greater than the major version of the version that included this feature, it will send a `RegisterAppInterface` response with `success=true` and move forward with this proposal's feature.
```

![registeringApp_flow](../assets/proposals/0280-Adding-new-parameter-of-requiresAudioSupport-and-BluetoothDeviceAddress/registeringApp_flow.png)

1. The developer will set their `requiresAudioSupport` to true.
2. The library will send a `StartService` for the RPC service with a new param in the payload, `requiresAudioSupport`.
3. Core will receive the `StartService` for the RPC service:
3-1. If `requiresAudioSupport` was set to false or not set, it will continue in flow.
3-2. Core will check it's current audio connects (BT A2DP), if it has a supported audio connect it will continue in flow.
3-3. If there is no Audio support, core will check if it supports the auto connect BT function. If it does it will continue in flow.

> 3-4. If `requiresAudioSupport` is true and there is no supported audio methods and Core does not support the auto connect BT function, it will send a `StartServiceNAK` with the reason `No audio support available`.

Reviewer: Please fix the formatting of these sub-bullets.  See attached for how they're currently displayed.  These should each have their own line and be indented under Section 3.
![Screen Shot 2020-04-23 at 10 59 06 AM](https://user-images.githubusercontent.com/22600496/80114523-b09ac680-8551-11ea-8fc3-901b5f8a1deb.png)

Reviewer: This is not yet resolved.  Please see attached screenshots.  You may need to add line breaks to make sure the sub-bullets are on their own line and indented.
![Screen Shot 2020-04-27 at 9 12 50 AM](https://user-images.githubusercontent.com/22600496/80376276-94f03280-8867-11ea-9f9f-cb9a50f62107.png)

4. If it supports the auto connect BT function, a new param in the payload, autoBTCapability set to true, otherwise set to false. Then if core has continued, it will send a `StartServiceACK`.
5. The app receives the response to its StartService for the RPC service:
5-1. If the response was a `StartServiceNAK`, the app will shutdown.
5-2. If the response was a `StartServiceACK`, it will continue in flow.
5-3. If `requiresAudioSupport` was set to false or not set, it will continue in flow.
5-4. The library check it's current audio connects (BT A2DP), if it has a supported audio connect it will continue in flow.

> 5-5. If `autoBTCapability` was set to true, t will continue in flow.

Author: Please refer to the flow because the conditions have changed.
The protocol version is related but can be determined from the parameters.

> 5-5. If `autoBTCapability` was set to true, t will continue in flow.

Reviewer: @Kazuki-Sugimoto the proposal needs to be updated based on the Steering Committee's requested revisions, noted [here](https://github.com/smartdevicelink/sdl_evolution/issues/941#issuecomment-600367823).

> 5-6. If the response was a `StartServiceACK` and `requiresAudioSupport` was set to true and autoBTCapability was set to false, the app will shutdown.

Reviewer: Please fix the formatting of these sub-bullets, so they are each on their own line and indented under Section 5.

Reviewer: This is not yet resolved.  Please see comment above regarding line breaks.

> 5-6. If the response was a `StartServiceACK` and `requiresAudioSupport` was set to true and autoBTCapability was set to false, the app will shutdown.

Reviewer: It doesn't appear that these changes have been made.  Please advise.

> 6. The app will send its RegisterAppInterface which will include `requiresAudioSupport` and `bluetoothDeviceAddress` in `deviceInfo`.

Reviewer: It doesn't appear that these changes have been made. Please advise.

> 7. When core receives RegisterAppInterface, it sends response. Then it will send `OnAppRegistered` include `deviceInfo` to HMI.

Author: I don't think `success=false` response other than an error is needed at this point.
It is not necessary to check whether this function is supported because it was judged before this point.
I don't know if that is possible, but `isMediaApplication` is not checked in case the media app does not use audio or the non-media apps use audio.

> 7. When core receives RegisterAppInterface, it sends response. Then it will send `OnAppRegistered` include `deviceInfo` to HMI.

Reviewer: @Kazuki-Sugimoto the proposal needs to be updated based on the Steering Committee's requested revisions, noted [here](https://github.com/smartdevicelink/sdl_evolution/issues/941#issuecomment-600367823).

1. The developer will set their `requiresAudioSupport` to true.
2. The library will send a `StartService` for the RPC service with a new param in the payload, `requiresAudioSupport`.
3. Core will receive the `StartService` for the RPC service: 

    1. Core will check its current audio connects (BT A2DP), if it has a supported audio connect it will continue in flow.
    2. If there is no audio support, Core will check if it supports the auto connect BT feature. If it does, it will continue in flow.
    3. If there is no supported audio methods and Core does not support the auto connect BT feature, it will send a `StartServiceNAK` with the reason `No audio support available`.

4. If Core has continued, it will send a `StartServiceACK` with a new param `autoBTCapability` set to true.
5. The app receives the response to its `StartService` for the RPC service:

    1. If the response was a `StartServiceNAK` the app will shut down.
    2. If the response was a `StartServiceACK`, `requiresAudioSupport` was set to true, but the protocol version of the ACK is less than the major version of this feature, the app will shut down.

>     3. If the response was a `StartServiceACK`, `requiresAudioSupport` was set to true, the protocol version of the ACK is equal or greater than the major version of this feature, and the `autoBTCapability` flag is set to false, the app will check if audio support is available using the `MediaStreamingStatus` class. If it is available it will continue, if not it will shut down.

Reviewer: ```suggestion
    3. If the response was a `StartServiceACK`, `requiresAudioSupport` was set to true, the protocol version of the ACK is equal to or greater than the major version of this feature, and the `autoBTCapability` flag is set to false, the app will check if audio support is available using the `MediaStreamingStatus` class. If it is available it will continue, if not it will shut down.
```

>     4. If the response was a `StartServiceACK`, `requiresAudioSupport` was set to true, the protocol version of the ACK is equal or greater than the major version of this feature, and the `autoBTCapability` flag is set to true, the app will continue.

Reviewer: ```suggestion
    4. If the response was a `StartServiceACK`, `requiresAudioSupport` was set to true, the protocol version of the ACK is equal to or greater than the major version of this feature, and the `autoBTCapability` flag is set to true, the app will continue.
```

6. The app will send its `RegisterAppInterface` which will include the `hmiTypes` and `isMediaApplication` flag.
7. Core will receive the `RegisterAppInterface`: 

    1. If the `requiresAudioSupport` flag was not included in the `StartService`, `isMediaApplication` is set to true in the RAI and the protocol version for the session is less than the major version of the version that included this feature, it will send a `RegisterAppInterface` response with `success=false` and deny the app's registration.

>     2. If the `requiresAudioSupport` flag was not included in the `StartService`, `isMediaApplication` is set to true in the RAI and the protocol version for the session is equal to or greater than the major version of the version that included this feature, it will send a `RegisterAppInterface response` with `success=true` but not move forward with this proposal's feature.

Reviewer: ```suggestion
    2. If the `requiresAudioSupport` flag was not included in the `StartService`, `isMediaApplication` is set to true in the RAI and the protocol version for the session is equal to or greater than the major version of the version that included this feature, it will send a `RegisterAppInterface` response with `success=true` but not move forward with this proposal's feature.
```

</struct>
```

### Launch the app

By using this function, the HU is able to perform  BT automatic connection based on the `bluetoothDeviceAddress` which is notified when the app that requests audio (`requiresAudioSupport` is set to true) is launched. 
It can also display a message that prompts BT connection to the user.

## Potential downsides

Due to the complexity of the flow, the developer must do the implementation carefully.

## Impact on existing code

This proposal requires a major version change.

> Since new parameters are added, Core, iOS, Java Suite, RPC, and HMI are affected.

Reviewer: Protocol should also be added as an impacted platform here.

## Alternatives considered

> The flow shown in Proposed solution is complicated, so the conditions have been arranged.

Reviewer: I'm having trouble understanding the meaning of this sentence, specifically "so the conditions have been arranged."  Can you please clarify and re-word?

You could update to "The flow in the `Proposed solution` section is complicated, so the author has also considered the simplified solution below." if this is your intended meaning.

Author: Sorry for the poor English translation. What I intended is as you wrote.

### Change the app registration flow

![registeringApp_flow2](../assets/proposals/0280-Adding-new-parameter-of-requiresAudioSupport-and-BluetoothDeviceAddress/registeringApp_flow2.png)

1. The developer will set their `requiresAudioSupport` to true.
2. The library will send a `StartService` for the RPC service with a new param in the payload, `requiresAudioSupport`.
3. Core will receive the `StartService` for the RPC service:
    i. If `requiresAudioSupport` was set to false or not set, it will continue in flow.
    ii. Core will check its current audio connects (BT A2DP), if it has a supported audio connect it will continue in flow.

>     iii. If there is no Audio support, Core will check if it supports the auto connect BT function. If it does it will continue in flow.

Reviewer: ```suggestion
    iii. If there is no audio support, Core will check if it supports the auto connect BT function. If it does, it will continue in flow.
```

>     iv. If `requiresAudioSupport` is true and there is no supported audio methods and Core does not support the auto connect BT function, it will send a `StartServiceNAK` with the reason `No audio support available`.

Reviewer: Please add line breaks to fix the formatting, so each item in the list is on its own line.

Author: Added line breaks.

4. If it supports the auto connect BT function, a new param in the payload of `StartServiceACK`, `autoBTCapability` is set to true, otherwise set to false. Then if Core has continued, it will send a `StartServiceACK`.
5. The app receives the response to its `StartService` for the RPC service:
    i. If the response was a `StartServiceNAK`, the app will shut down.
    ii. If the response was a `StartServiceACK`, it will continue in flow.
    iii. If `requiresAudioSupport` was set to false or not set, it will continue in flow.

>     iv. The library check it's current audio connects (BT A2DP), if it has a supported audio connect it will continue in flow.

Reviewer: ```suggestion
    iv. The library checks its current audio connects (BT A2DP), if it has a supported audio connect, it will continue in flow.
```

    v. If `autoBTCapability` was set to true, it will continue in flow.

>     vi. If the response was a `StartServiceACK` and `requiresAudioSupport` was set to true and autoBTCapability was set to false, the app will shutdown.

Reviewer: Please add line breaks to fix the formatting, so each item in the list is on its own line.

Author: Added line breaks.

>     vi. If the response was a `StartServiceACK` and `requiresAudioSupport` was set to true and autoBTCapability was set to false, the app will shutdown.

Reviewer: ```suggestion
    vi. If the response was a `StartServiceACK` and `requiresAudioSupport` was set to true and `autoBTCapability` was set to false, the app will shut down.
```

> 6. The app will send its RegisterAppInterface which will include `requiresAudioSupport` and `bluetoothDeviceAddress` in `deviceInfo`.

Reviewer: ```suggestion
6. The app will send its `RegisterAppInterface` which will include `requiresAudioSupport` and `bluetoothDeviceAddress` in `deviceInfo`.
```

> 7. When core receives RegisterAppInterface, it sends response. Then it will send `OnAppRegistered` include `deviceInfo` to HMI.

Reviewer: ```suggestion
7. When Core receives `RegisterAppInterface`, it sends response. Then it will send `OnAppRegistered` including `deviceInfo` to HMI.
```

#### Preconditions

As demonstrated in the flow, the following conditions are required to use this function.
If the SDL app or HU cannot use this function, it will perform the same operation as before.

|conditions|details|
|---|---|
|The app/HMI must have the version that supports this function|App side: `requiresAudioSupport` of `RegisterAppInterface` must be set to true in order to use this function.<br>HMI side: `autoBTCapability` of `StartServiceACK` must be set to true in order to use this function.|
|App must request audio|In case of media app, `requestAudioSupport` is set to true as default. However, developers can set `requiresAudioSupport` to true regardless of default settings.|
|HU must have BT automatic connection function|`autoBTConnect` of` HMICapabilities` must be set to true.|

#### Add parameters

> Only the differences from the proposal are shown below.

Reviewer: ```suggestion
Only the differences from the `Proposed solution` section are shown below.
```
